crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/bootstrap_task.rs

```diff
@@ -0,0 +1,756 @@
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_query_agent_limit_is_trusted() {
+        // The code expects a max of 8 agents, send more
+        let agents = std::iter::repeat_with(|| fixt!(AgentInfoSigned))
+            .take(30)
+            .collect::<Vec<_>>();
+        let (test_sender, mut host_stub, _) = setup(
+            DummySpaceInternalImpl::new(HashSet::new()),
+            agents,
+            2,
+            false,
+        )
+        .await;
+
+        let evt = host_stub.next_event(Duration::from_secs(5)).await;
+
+        // TODO This should probably be an error rather than being forwarded to the host
+        assert_eq!(30, evt.peer_data.len());
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_local_agents_in_response_are_filtered() {
+        // The code expects a max of 8 agents, send more
+        let agents = std::iter::repeat_with(|| fixt!(AgentInfoSigned))
+            .take(10)
+            .collect::<Vec<_>>();
+
+        let local_agents = agents.iter().take(3).cloned().collect::<HashSet<_>>();
+
+        let (test_sender, mut host_stub, _) =
+            setup(DummySpaceInternalImpl::new(local_agents), agents, 2, false).await;
+
+        let evt = host_stub.next_event(Duration::from_secs(5)).await;
+
+        assert_eq!(7, evt.peer_data.len());
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_handles_errors_sending_to_host() {
+        let agents = vec![fixt!(AgentInfoSigned)];
+        let (test_sender, mut host_stub, _) =
+            setup(DummySpaceInternalImpl::new(HashSet::new()), agents, 2, true).await;
+
+        // Ask the host to respond with an error on each call, then wait a while and clear the flag
+        host_stub.respond_with_error.store(true, Ordering::SeqCst);
+        tokio::time::sleep(Duration::from_millis(10)).await;
+        host_stub.respond_with_error.store(false, Ordering::SeqCst);
+
+        // Now expect to receive from the task as usual
+        let receives = Arc::new(AtomicUsize::new(0));
+        tokio::time::timeout(Duration::from_secs(30), {
+            let task_receives = receives.clone();
+            async move {
+                for _ in 0..3 {
+                    host_stub.next_event(Duration::from_secs(5)).await;
+                    task_receives.fetch_add(1, Ordering::SeqCst);
+                }
+            }
+        })
+        .await
+        .unwrap();
+
+        assert_eq!(
+            3,
+            receives.load(Ordering::SeqCst),
+            "Expected 3 calls to have succeeded"
+        );
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_shuts_down_if_host_closes() {
```

Comment:
Thought of another test case after putting this up for review. What happens if the channel we use to the host closes? Turns out that the task keep running and polling the bootstrap server even though it can't persist the results.

The task code has been updated to make this test pass.